### ISSUE TITLE
It should be possible to customize Docker image names created with azd deploy

### ISSUE THREAD

**Issue by christiannagel** [Azure/azure-dev#3335]
azd version 1.5.1 (commit 3856d1e98281683b8d112e222c0a7c7b3e148e96)

Using `azd deploy` or `azd up` with .NET Aspire projects, docker images with such names are generated:
azd-deploy-bot-1707570986
azd-deploy-bot-1707571578

Every deployment creates a new repository with ACR, all with the `latest` tag.

I would like to create images names such as `codebreaker-bot`, and create a tag with the version number, and just have the latest build tagged with the `latest` tag.

I tried to use the `ContainerRepository` element to name the image in the csproj file - but likely azd uses the command-line with `dotnet publish` which overwrites the project file settings. Setting the image name and tags in other ways than using the project file would also be ok, but it should be possible to customize this.

Using
-  azd version 1.5.1 (commit 3856d1e98281683b8d112e222c0a7c7b3e148e96)
- .NET Aspire  8.0.0-preview.2.23619.3/8.0.100
- Visual Studio 17.9.0 Preview 5

**Comment by baronfel** [comment]
big +1 to this - the `ContainerRepository` should be left to the user, and tooling should set `ContainerImageTag` or `ContainerImageTags` to whatever it needs to uniquely identify a container image/revision.

**Comment by timheuer** [comment]
I would personally like `azd` to publish to one repository with tags...having the latest deployment be a tag + `latest`

**Comment by rajeshkamal5050** [comment]
@davidfowl @vhvb1989 @mitchdenny @ellismg moved it as a fit-n-finish item post Preview 5. Please let me know if you think otherwise.

**Comment by vicperdana** [comment]
@rajeshkamal5050 would this [question](https://github.com/dotnet/aspire/discussions/4002) fall in the same bucket?

**Comment by joseph-hungerman** [comment]
Whatever happened with this? I'm struggling with this issue now. I have like 40 repositories in my container app when I just want one repo with the rev history and one with the latest tag

**Comment by vhvb1989** [comment]
@ellismg , IIRC, you had the fix for this already in a branch. Can we have a PR?

**Comment by paulyuk** [comment]
I like how this fix enables more languages, but it does not address the original need which is to be able to set and override the container image tag.

I'd like to be able to do something like this:
```

          # Using a descriptive tag that combines branch, run number, and short SHA
          # This is the best practice for traceability and immutability.
          COMMIT_SHA_SHORT=${GITHUB_SHA::7}
          **IMAGE_TAG="${{ github.ref_name }}-${{ github.run_number }}-${COMMIT_SHA_SHORT}"**
          export AZURE_CONTAINER_IMAGE_TAG=${IMAGE_TAG}

          echo "AZURE_CONTAINER_IMAGE_TAG=${IMAGE_TAG}" >> $GITHUB_ENV
          echo "Generated image tag: ${IMAGE_TAG}"

..

            echo "🚀 Main branch - deploying application"            # Set container image tag before running azd up
            **azd env set AZURE_CONTAINER_IMAGE_TAG $AZURE_CONTAINER_IMAGE_TAG**
            echo "✅ Set container image tag to: $AZURE_CONTAINER_IMAGE_TAG"
            
            # Run azd up with the correct environment name
            if [ "$AZD_DEBUG" = "true" ]; then
              echo "🐞 Debug mode enabled for AZD"
              azd up --no-prompt --debug
            else
              echo "🚀 Running AZD in normal mode"
              azd up --no-prompt
            fi
```

Could we please reopen?  @kristenwomack @spboyer

**Comment by spboyer** [comment]
Reopened - @kristenwomack @rajeshkamal5050

**Comment by rajeshkamal5050** [comment]
@weikanglim can you take a look at how to enable customizing/overriding docker image tags. Also refer to @paulyuk comment above for the suggestion.

**Comment by weikanglim** [comment]
@paulyuk Thanks for the helpful example! I believe what you're looking for is achievable today with the following azure.yaml configuration:

```yaml
services:
  web:
    project: react-fluent
    language: js
    host: containerapp
    docker:
      image: ${AZURE_CONTAINER_IMAGE_TAG}
```

Using the above configuration, azd will use the environment variable `AZURE_CONTAINER_IMAGE_TAG` to tag the generated image if set (which can be useful for your CI use case), or default back to the azd auto-generated image tag if unset.

**Comment by spboyer** [comment]
@weikanglim this may be the case for all languages EXCEPT .NET Aspire where the user does not have access to the azure.yaml file. This is an internal process and there is no inherent way to set this.

**Comment by joseph-hungerman** [comment]
I may be mistaken but running azd infra synth gives you access to the yaml at the solution root

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by vhvb1989** [comment]
That is correct @joseph-hungerman . However, `infra synth` produces either bicep or YAML to define the Azure Container App or other Azure Resources.  It does not provide a way to customize the name of the container.  AZD picks the name of the container when it runs `dotnet publish` .

We can make AZD to supports a way to control the name. But I think there should be a way to control the name directly from the AppHost. 

@davidfowl Can we have an API in the AppHost to control the name?  That way all publishers would honor that as the source of truth .  It would be a shame if changing from one publisher to another causes containers to be named different. right?

**Comment by davidfowl** [comment]
Yes, I think this should be controlled by an apphost api and I think using parameters (as you suggested) is the right thing to do.

**Comment by snavarropino** [comment]
I am a bit confused.  Should setting image in yaml work?  Is not working for me

```yml
services:
  web:
    project: react-fluent
    language: js
    host: containerapp
    docker:
      image: ${AZURE_CONTAINER_IMAGE_TAG}
```

**Comment by vhvb1989** [comment]
@snavarropino it does not work for Aspire projects. It does work for non-aspire (NET projects).

For Aspire projects, there should be (eventually) one API like `.withContainerName(....)` which you could use for your Project.  Do you have an issue in Aspire side which we could track @davidfowl

**Comment by rajeshkamal5050** [comment]
> [@snavarropino](https://github.com/snavarropino) it does not work for Aspire projects. It does work for non-aspire (NET projects).
> 
> For Aspire projects, there should be (eventually) one API like `.withContainerName(....)` which you could use for your Project. Do you have an issue in Aspire side which we could track [@davidfowl](https://github.com/davidfowl)

Works for non-Aspire projects. For Aspire, support needs to be added in the Aspire AppModel. @vhvb1989 can you create an iissue in Aspire for the same and close this?

**Comment by rajeshkamal5050** [comment]
@vhvb1989 is this the aspire dependency which fixes it? 
- https://github.com/dotnet/aspire/issues/1855

**Comment by vhvb1989** [comment]
> [@vhvb1989](https://github.com/vhvb1989) is this the aspire dependency which fixes it?
> 
> * [Support setting a specific container name for container resources in AppHost projects dotnet/aspire#1855](https://github.com/dotnet/aspire/issues/1855)

It was never implemented. That's what it would fix it, yes

**Comment by vhvb1989** [comment]
Keeping this enhancement open for review against the new model. Custom image naming depends on the AppHost/Aspire publishing contract and should be evaluated as part of the complete extension deployment workflow in #9482. Overall roadmap: #9479.
